### ISSUE TITLE
User can't launch DrRacket due to problem in `key-translate`

### ISSUE THREAD

**Issue by soegaard** [racket/gui#330]
From https://www.reddit.com/r/Racket/comments/1dwusoe/help_i_cant_use_the_drracket_app/

```
[lambduli](https://www.reddit.com/user/lambduli)
Hey everyone,

Today I installed the latest version `8.13` and it doesn't seem to work.

Here's what my setup is and what I tried:

I am on Apple MacBook Air with M1 chip. The latest OS.

The app from the launcher doesn't work. The icon bounces but then stops.

When I tried running it from the terminal I got this error:

integer->char: contract violation
  expected: valid-unicode-scalar-value?
  given: 55349
  context...:
   /Applications/Racket v8.13/share/pkgs/gui-lib/mred/private/wx/cocoa/key-translate.rkt:454:0: key-translate
   /Applications/Racket v8.13/share/pkgs/gui-lib/mred/private/wx/cocoa/key-translate.rkt:552:0
   body of "/Applications/Racket v8.13/share/pkgs/gui-lib/mred/private/wx/cocoa/key-translate.rkt"
   body of "/Applications/Racket v8.13/share/pkgs/gui-lib/mred/private/wx/platform.rkt"
   
   EDIT
   So it must be related to my layout. I have a custom keyboard layout with bunch of special symbols on it. When I switch to a simpler layout I can open DrRacket just fine. I think this might be a bug. I should probably open an issue.
   ```
 
 The last line of `key-translate` is:
 ```
 (integer->char (ptr-ref output-chars _UniChar i))
 ```
 
The number 55349 from the error message falls outside the ranges accepted by `integer->char`.
The accepted ranges are from 0 to D7FF and from E000 to 10FFFF.

The documentation from Apple says an array of UniChar is returned.
UniChar stands for "UTF-16 code units".

A search for 55349 and UTF 16 confirms that 55349 is not a valid UTF 16 character.

Suggestion: We return some dummy value like `#\space` for integers from D800 to DFFF.

**Comment by lambduli** [comment]
Here is my keyboard layout that seems to be causing the issues.
[Code.keylayout.zip](https://github.com/user-attachments/files/16117409/Code.keylayout.zip)
I had to zip it for GitHub. I will add screenshots in the next one.

**Comment by lambduli** [comment]
Screenshots of my layout.

No special keys pressed:
<img width="752" alt="Screenshot 2024-07-06 at 23 40 58" src="https://github.com/racket/gui/assets/6469748/5363dfbc-6fb6-494c-9c1c-e6095ec5724f">


Shift:
<img width="752" alt="Screenshot 2024-07-06 at 23 41 06" src="https://github.com/racket/gui/assets/6469748/b80ee3a7-3d06-4784-98fc-ba5d456f6c1b">



Option key:
<img width="752" alt="Screenshot 2024-07-06 at 23 41 13" src="https://github.com/racket/gui/assets/6469748/8b83281c-b669-4792-8e8e-ad5c08961c9e">


Option key + Shift key:
<img width="752" alt="Screenshot 2024-07-06 at 23 41 21" src="https://github.com/racket/gui/assets/6469748/0670fc34-ebf5-4ff4-aa82-60b7419982d9">


Caps lock key:
<img width="752" alt="Screenshot 2024-07-06 at 23 41 32" src="https://github.com/racket/gui/assets/6469748/4be4b80f-fe17-4288-aa7a-05b6b84f6b8b">


Caps lock + Shift:
<img width="752" alt="Screenshot 2024-07-06 at 23 41 38" src="https://github.com/racket/gui/assets/6469748/54e1e0f1-f2cb-418d-9006-d002b86e5865">



Control key + option key:
<img width="752" alt="Screenshot 2024-07-06 at 23 41 51" src="https://github.com/racket/gui/assets/6469748/9a8eb80e-650d-47c6-9a28-d4352b41bcb4">

**Comment by lambduli** [comment]
I am on Apple MacBook Air Sonoma 14.5

**Comment by soegaard** [comment]
Interesting layout - you've got to be a mathematician?

In the file `gui-lib/mred/private/wx/cocoa/key-translate.rkt` at line 495-496
we have:

```
(list->string (for/list ([i (in-range n)]) 
                       (integer->char (ptr-ref output-chars _UniChar i)))))
```

The problem is that your keyboard layout results in an integer (utf-16 character) 55349 
outside the valid range.

Do you time to test the following? We simply ignore integers in the non-valid range:

```
(list->string (for/list ([i (in-range n)] #:when (not (<= #xD800 (ptr-ref output-chars _UniChar i) #xDFFF))) 
                       (integer->char (ptr-ref output-chars _UniChar i)))))
```

**Comment by lambduli** [comment]
I do stuff around type systems and type theory. I like to use these because it makes the formulae and all that more readable to me (`m₁` reads much better than `m1` or `m_1` and such).

I'd be happy to test anything. I can't find the file in my installation so I at least tried to run the second expression but it gives me an error: `?: bad syntax in: #:when`

**Comment by lambduli** [comment]
Oh, sorry — found it. I just had to look up the error again.

**Comment by soegaard** [comment]
@lambduli 
You can find the file with this command:

```
mdfind key-translate.rkt | grep Applications
```

On my system the location is:
```
/Applications/Racket v8.12/share/pkgs/gui-lib/mred/private/wx/cocoa/key-translate.rkt
```
Open the file in a text editor and replace:
```
(list->string (for/list ([i (in-range n)]) 
                       (integer->char (ptr-ref output-chars _UniChar i)))))
```
with
```
(list->string (for/list ([i (in-range n)] #:when (not (<= #xD800 (ptr-ref output-chars _UniChar i) #xDFFF))) 
                       (integer->char (ptr-ref output-chars _UniChar i)))))
```

**Comment by lambduli** [comment]
It seems that it worked! I can open it from the launcher without having to switch to my simpler layout.

**Comment by lambduli** [comment]
> @lambduli You can find the file with this command:
> 
> ```
> mdfind key-translate.rkt | grep Applications
> ```
> 
> On my system the location is:
> 
> ```
> /Applications/Racket v8.12/share/pkgs/gui-lib/mred/private/wx/cocoa/key-translate.rkt
> ```
> 
> Open the file in a text editor and replace:
> 
> ```
> (list->string (for/list ([i (in-range n)]) 
>                        (integer->char (ptr-ref output-chars _UniChar i)))))
> ```
> 
> with
> 
> ```
> (list->string (for/list ([i (in-range n)] #:when (not (<= #xD800 (ptr-ref output-chars _UniChar i) #xDFFF))) 
>                        (integer->char (ptr-ref output-chars _UniChar i)))))
> ```

Yeah you're right, I think my brain wasn't raining properly (it's quite late here). Thanks for the patience.

**Comment by soegaard** [comment]
> It seems that it worked! I can open it from the launcher without having to switch to my simpler layout.
> Yeah you're right, I think my brain wasn't braining properly for a minute (it's quite late here). 
Great! Ditto (same time zone).

**Comment by lambduli** [comment]
Thanks for your help! This was a nice experience.

**Comment by soegaard** [comment]
@lambduli Thanks for the help.

I have committed a PR.
The Github Action failed - not due to the PR - but because
the Ubuntu version was bumped and now uses a newer glibc
than expected by the build script (at least that's my current analysis).

**Comment by mflatt** [comment]
Fixed by #334
